tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c

```diff
@@ -721,43 +866,178 @@ static void az_ulib_ipc_init_succeed(void** state)
+  assert_int_equal(result, AZ_OK);
+
+  /// cleanup
+}
+
+/* If there is published interface, the az_ulib_ipc_deinit shall return AZ_ERROR_ULIB_BUSY. */
+static void az_ulib_ipc_deinit_with_published_interface_failed(void** state)
+{
+  /// arrange
+  (void)state;
+  az_ulib_ipc_interface_handle ipc_query_handle;
+  init_ipc_and_publish_interfaces();
+
+  /// act
+  az_result result = az_ulib_ipc_deinit();
+
+  /// assert
+  assert_int_equal(result, AZ_ERROR_ULIB_BUSY);
+  assert_int_equal(
+      az_ulib_ipc_try_get_interface(
+          AZ_SPAN_EMPTY,
+          AZ_SPAN_FROM_STR(IPC_1_PACKAGE_NAME),
+          IPC_1_PACKAGE_VERSION,
+          AZ_SPAN_FROM_STR(QUERY_1_INTERFACE_NAME),
+          QUERY_1_INTERFACE_VERSION,
+          &ipc_query_handle),
+      AZ_OK);
+
+  /// cleanup
+  assert_int_equal(az_ulib_ipc_release_interface(ipc_query_handle), AZ_OK);
+  unpublish_interfaces_and_deinit_ipc();
+}
+
+/* If there is instances of the interface, the az_ulib_ipc_deinit shall return AZ_ERROR_ULIB_BUSY.
+ */
+static void az_ulib_ipc_deinit_with_instace_failed(void** state)
+{
+  /// arrange
+  (void)state;
+  az_ulib_ipc_interface_handle ipc_query_handle;
+  init_ipc_and_publish_interfaces();
+  az_ulib_ipc_interface_handle interface_handle;
+  assert_int_equal(
+      az_ulib_ipc_try_get_interface(
+          AZ_SPAN_EMPTY,
+          AZ_SPAN_FROM_STR(MY_PACKAGE_A_NAME),
+          MY_PACKAGE_1_VERSION,
+          AZ_SPAN_FROM_STR(MY_INTERFACE_1_NAME),
+          MY_INTERFACE_123_VERSION,
+          &interface_handle),
+      AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_a_1_1_123_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_b_1_1_123_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_c_1_1_123_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_a_2_1_123_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_a_1_2_123_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_a_1_1_200_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+  assert_int_equal(az_ulib_test_my_interface_a_1_3_123_unpublish(AZ_ULIB_NO_WAIT), AZ_OK);
+
+  /// act
+  az_result result = az_ulib_ipc_deinit();
+
+  /// assert
+  assert_int_equal(result, AZ_ERROR_ULIB_BUSY);
+  assert_int_equal(
+      az_ulib_ipc_try_get_interface(
+          AZ_SPAN_EMPTY,
+          AZ_SPAN_FROM_STR(IPC_1_PACKAGE_NAME),
+          IPC_1_PACKAGE_VERSION,
+          AZ_SPAN_FROM_STR(QUERY_1_INTERFACE_NAME),
+          QUERY_1_INTERFACE_VERSION,
+          &ipc_query_handle),
+      AZ_OK);
+
+  /// cleanup
+  assert_int_equal(az_ulib_ipc_release_interface(ipc_query_handle), AZ_OK);
+  assert_int_equal(az_ulib_ipc_release_interface(interface_handle), AZ_OK);
+  assert_int_equal(az_ulib_ipc_deinit(), AZ_OK);
+}
+
+/* The az_ulib_ipc_publish shall store the descriptor published in the IPC. The az_ulib_ipc_publish
  * shall be thread safe. */
```

Comment:
>shall be thread safe. [](http://example.com/codeflow?start=3&length=21)

are you testing thread safety here?

Reply:
Yes, and in the e2e test as well. Here we mock the lock to do that.
```
  assert_int_equal(g_lock_diff, 0);
  assert_int_equal(g_count_acquire, 6);
```